### PR TITLE
[FEATURE] enable css-vars and skeleton build

### PR DESCRIPTION
**Pull request by petermuessig**
Enable CSS variables support:
https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties

can be enabled via compiler option `cssVars`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -151,6 +151,12 @@ Options for the [less](http://lesscss.org) compiler (`tree.toCss`).
 
 **Note:** The options `sourceMap` and `cleancss` are not supported. An exception will be thrown when used.
 
+###### cssVars
```
Comment (matz3) [inline README.md]: As the option is not relevant for the less compiler, I would rather have it next to the other custom options such as "rtl" or "rootPaths".

File: lib/index.js [hunk 2]
```diff
@@ -118,6 +117,15 @@ Builder.prototype.cacheTheme = function(result) {
 	});
 };
 
+/**
+ * Creates a themebuild
+ * @param {object} options
+ * @param {object} options.compiler compiler object as passed to less
+ * @param {boolean} options.compiler.cssVars whether or not to enable css variables output
```
Comment (matz3) [inline lib/index.js]: `options.cssVars` (see comment in readme)

File: lib/index.js [hunk 3]
```diff
@@ -238,16 +247,50 @@ Builder.prototype.build = function(options) {
...
 			// also compile rtl-version if requested
+			let oRTL;
 			if (options.rtl) {
-				result.cssRtl = tree.toCSS(assign(options.compiler, {
-					plugins: [new RTLPlugin()]
+				const RTLPlugin = require("./plugin/rtl");
+				oRTL = new RTLPlugin();
+			}
+
+			if (oRTL) {
```
Comment (matz3) [inline lib/index.js]: Why not including this into the "if" above?

File: lib/index.js [hunk 4]
```diff
@@ -238,16 +247,50 @@ Builder.prototype.build = function(options) {
...
+						plugins: [oCSSVariablesCollectorRTL, oRTL]
+					}));
+				}
+				return new Promise(function(resolve, reject) {
+					// generate the CSS variables content out of the less variables
+					parser.parse(result.cssVars.source, function(err, tree) {
+						if (err) {
+							reject(err);
+						} else {
+							result.cssVars.vars = tree.toCSS(assign({}, options.compiler));
```
Comment (tobiasso85) [inline lib/index.js]: would it make sense to rename `vars` to `cssVariables` since it will be the content of the variables.css?

**Comment by petermuessig** [comment]
Ah - ok - I will also incorporate your feedback. For now I introduced a library scope for the dynamic variables.  This ensures no duplicates across the libraries which could happen in theory. A scoping per theme is not needed as they are not mixed.

File: README.md (hunk 1 above)
Comment (petermuessig) [reply README.md]: Ok

File: lib/index.js (hunk 2 above)
Comment (petermuessig) [reply lib/index.js]: Ok

File: lib/index.js (hunk 4 above)
Comment (petermuessig) [reply lib/index.js]: I adopt `vars` to `cssVariables` and `source` to `lessVariables`

File: lib/index.js (hunk 3 above)
Comment (petermuessig) [reply lib/index.js]: I did this to explicitly make clear that I run the RTL build only if the RTL plugin is being created. The same check exists also a bit more down.

**Comment by petermuessig** [comment]
Comments are incorporated!

**Comment by petermuessig** [comment]
One last thing is to avoid the redundancies in the CSS variables list - let's discuss this briefly...

File: lib/index.js [hunk 5]
```diff
@@ -359,6 +426,15 @@ Builder.prototype.build = function(options) {
 				results.embeddedCompare.cssRtl += applyScope(results.embeddedCompare.cssRtl, results.embedded.cssRtl, true);
 			}
 
+			if (options.cssVars) {
+				// results.embeddedCompare.cssVars.lessVariables += applyScope(results.embeddedCompare.cssVars.lessVariables, results.embedded.cssVars.lessVariables);
```
Comment (matz3) [inline lib/index.js]: Can be removed?

File: lib/plugin/css-variables-collector.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+		// reset overrides
+		less.tree.Variable.prototype.genCSS = undefined;
+		less.tree.Quoted.prototype.genCSS = fnQuotedgenCSS;
+
+		return css;
+	},
+
+	run(root) {
+		// console.log("run", root);
```
Comment (matz3) [inline lib/plugin/css-variables-collector.js]: Remove

File: lib/plugin/css-variables-collector.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		return this.native.visit(root);
+	},
+
+	visitOperation(node, visitArgs) {
+		// console.log("visitOperation", node, visitArgs);
```
Comment (matz3) [inline lib/plugin/css-variables-collector.js]: Remove

File: lib/index.js [hunk 8]
```diff
@@ -194,6 +202,12 @@ Builder.prototype.build = function(options) {
 			parserOptions.filename = config.localPath;
 		}
 
+		// inject the library name as prefix (e.g. "sap.m" as "_sapM")
```
Comment (matz3) [inline lib/index.js]: It's `_sap_m`, right? (not _sapM)

File: lib/index.js [hunk 9]
```diff
@@ -194,6 +202,12 @@ Builder.prototype.build = function(options) {
...
+		if (options.library && typeof options.library.name === "string") {
+			const libName = config.libName = options.library.name;
+			config.prefix = "_" + libName.replace(/\./g, "_");
```
Comment (matz3) [inline lib/index.js]: Maybe better add another `_` at the end to ensure that there is a separation?

**Pull request merged**